nuxeo-runtime/nuxeo-runtime-test/src/main/java/org/nuxeo/runtime/test/RuntimeHarnessImpl.java

```diff
@@ -0,0 +1,478 @@
+import org.nuxeo.osgi.JarBundleFile;
+import org.nuxeo.osgi.OSGiAdapter;
+import org.nuxeo.osgi.SystemBundle;
+import org.nuxeo.osgi.SystemBundleFile;
+import org.nuxeo.osgi.application.StandaloneBundleLoader;
+import org.nuxeo.runtime.RuntimeServiceException;
+import org.nuxeo.runtime.api.Framework;
+import org.nuxeo.runtime.model.Extension;
+import org.nuxeo.runtime.model.RegistrationInfo;
+import org.nuxeo.runtime.model.RuntimeContext;
+import org.nuxeo.runtime.model.StreamRef;
+import org.nuxeo.runtime.model.URLStreamRef;
+import org.nuxeo.runtime.model.impl.DefaultRuntimeContext;
+import org.nuxeo.runtime.osgi.OSGiRuntimeContext;
+import org.nuxeo.runtime.osgi.OSGiRuntimeService;
+import org.nuxeo.runtime.test.protocols.inline.InlineURLFactory;
+import org.nuxeo.runtime.test.runner.RuntimeHarness;
+import org.nuxeo.runtime.test.runner.TargetExtensions;
+import org.nuxeo.runtime.transaction.TransactionHelper;
+import org.osgi.framework.Bundle;
+import org.osgi.framework.FrameworkEvent;
+
+import io.github.lukehutch.fastclasspathscanner.FastClasspathScanner;
+
+public final class RuntimeHarnessImpl implements RuntimeHarness {
+
+    protected static final Log log = LogFactory.getLog(RuntimeHarnessImpl.class);
+
+    protected static URL[] introspectClasspath(ClassLoader loader) {
+        return new FastClasspathScanner().getUniqueClasspathElements().stream().map(file -> {
+            try {
+                return file.toURI().toURL();
+            } catch (MalformedURLException cause) {
+                throw new Error("Could not get URL from " + file, cause);
+            }
+        }).toArray(URL[]::new);
+    }
+
+    protected StandaloneBundleLoader bundleLoader;
+
+    protected Map<String, BundleFile> bundles;
+
+    protected boolean frameworkStarted;
+
+    protected OSGiAdapter osgi;
+
+    protected Set<URI> readUris;
+
+    protected OSGiRuntimeService runtime;
+
+    protected Bundle runtimeBundle;
+
+    protected final TargetResourceLocator targetResourceLocator;
+
+    protected URL[] urls; // classpath urls, used for bundles lookup
+
+    protected final List<WorkingDirectoryConfigurator> wdConfigs;
+
+    protected File workingDir;
+
+    public RuntimeHarnessImpl(Class<?> clazz) {
+        targetResourceLocator = new TargetResourceLocator(clazz);
+        wdConfigs = new ArrayList<>();
+    }
+
+    @Override
+    public void addWorkingDirectoryConfigurator(WorkingDirectoryConfigurator config) {
+        wdConfigs.add(config);
+    }
+
+    @Override
+    public void deployBundle(String name) throws Exception {
+        // install only if not yet installed
+        Bundle bundle = bundleLoader.getOSGi().getRegistry().getBundle(name);
+        if (bundle == null) {
+            BundleFile bundleFile = lookupBundle(name);
+            bundleLoader.loadBundle(bundleFile);
+            bundleLoader.installBundle(bundleFile);
+            bundle = bundleLoader.getOSGi().getRegistry().getBundle(name);
+        }
```

Comment:
Could add an `INFO` log to tell the bundle wasn't install ?

Reply:
Logged.